PLT-7404 Return viewed at times in view channel API response

jwilander opened the pull request:
#### Summary
I used a transaction to select the LastPostAt times out of the database and then update the ChannelMembers. This might have perf implications because it adds a read on a hot route so if either of you can think of a better method that doesn't require an extra read let me know

API reference update: mattermost/mattermost-api-reference#288

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-7404

ccbrown commented:
This API change just feels wrong since it doesn't actually make it possible to use the API in a race-free way.

But as far as your transaction goes, you're basically doing the same read twice: Once in your explicit SELECT, and again in the `UPDATE ChannelMembers, Channels` query. I think you can rewrite those such that there's not actually an additional read and the only performance impact is a round-trip between the read and write (Avoid the introduction of a new race condition with a condition to make sure the update doesn't move the columns backwards.).

Then if we really wanted to optimize this further, maybe we could use a stored procedure?

jwilander commented:
If you're strongly opinionated about this API change, feel free to take the ticket. If you're satisfied with this solution then let me know and I'll continue work on it.

As for the transaction, you're right. I can just select all the values needed for the update in the explicit select

ccbrown commented:
Yeah, sorry, I'm not opinionated strongly enough to take it myself. This is fine. 👍

jwilander commented:
Updated the transaction, take a look. For the race condition you mentioned about going backwards, what case do you see that happening in? I could make the update SQL query more complicated to make sure the new time is ahead of the old time but I'd rather avoid it since it's already rather complex

ccbrown commented:
I was thinking about the case where request A performs the SELECT, request B both begins and commits a transaction, then request A performs the UPDATE. The transaction won't do anything to stop request A from effectively rolling back request B. As-is, and we're using the default "repeatable read" isolation level, I don't think the transaction does anything at all for us. So I was thinking a race condition could be avoided by removing the transaction and using a sort of optimistic locking strategy instead.

jwilander commented:
I guess transactions weren't the save-the-day entities I thought they were. I removed the transaction and changed the update to only use the new value if it's greater than the old one

crspeller commented:
@jwilander Let me check the performance implementations of this. Assigned to myself.

crspeller commented:
@jwilander Can you rebase this so that I can loadtest it.

crspeller commented:
Loadtest results:
```
Route: /channels/members/me/view
Total Hits: 69501
Error Rate: 0.04%
Max Response Time: 1886ms
Min Response Time: 13ms
Mean Response Time: 77.02ms
Median Response Time: 58.00ms
Inter Quartile Range: 37
```

Baseline (last daily):
```
Route: /channels/members/me/view
Total Hits: 69510
Error Rate: 0.06%
Max Response Time: 2081ms
Min Response Time: 1ms
Mean Response Time: 84.41ms
Median Response Time: 62.00ms
Inter Quartile Range: 57
```

If anything there is an improvement for this route.

ccbrown commented:
> Min Response Time: 1ms

@crspeller Are the errors included in this calculation? Surely the load tests didn't actually serve a valid respond in 1ms, right? (Just curious.)
